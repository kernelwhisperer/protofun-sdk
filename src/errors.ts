export class IndexerError extends Error {
  name = "IndexerError"
}
