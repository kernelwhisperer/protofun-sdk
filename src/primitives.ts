export type Timeframe = "Block" | "Minute" | "Hour" | "Day" | "Week"

export enum PriceUnit {
  ETH = "ETH",
  USD = "USD",
  GWEI = "Gwei",
}

export type Block = {
  baseFeePerGas: string // BigInt
  burnedFees: string // BigInt
  firstGasPrice: string // BigInt
  gasFees: string // BigInt
  gasUsed: string // BigInt
  id: string // bytes
  lastGasPrice: string // BigInt
  maxGasPrice: string // BigInt
  minGasPrice: string // BigInt
  minerTips: string // BigInt
  number: string // BigInt
  timestamp: string // BigInt
  // txns: Array<Txn>
  txnCount: number
}

export type SimpleBlock = Omit<Block, "txns" | "timestamp"> & {
  timestamp: string
}

export type Candle = {
  close: string
  high: string
  low: string
  open: string
  timestamp: string
}

export function isCandle(value: unknown): value is Candle {
  return typeof value === "object" && value !== null && "open" in value
}

export function isCandleArray(value: unknown[]): value is Candle[] {
  return Array.isArray(value) && value.length > 0 && isCandle(value[0])
}

export type QueryRequest = {
  /**
   * @default 1000
   */
  limit?: number
  priceUnit?: PriceUnit
  since?: string
  timeframe: Timeframe
  until?: string
}
export type QueryResult = Promise<Candle[] | SimpleBlock[]>
export type QueryFn = (request: QueryRequest) => QueryResult
