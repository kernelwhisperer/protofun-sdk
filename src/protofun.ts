export * from "./errors"
export * from "./metric-declarations"
export * from "./primitives"
