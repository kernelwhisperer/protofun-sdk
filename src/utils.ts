import { MetricId, ProtocolId, QueryFn, SubscribeFn, Timeframe } from "./protofun"

export async function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export const allTimeframes: Timeframe[] = ["Block", "Minute", "Hour", "Day", "Week"]

export type MetricFnsResult = {
  query: QueryFn
  subscribe: SubscribeFn
  supportedTimeframes: Timeframe[]
}

export async function loadMetricFns(
  protocolId: ProtocolId,
  metricId: MetricId,
  packagePath = "protofun/dist/metrics"
): Promise<MetricFnsResult> {
  const module = await import(`${packagePath}/${protocolId}/${metricId}`)
  const { default: query, supportedTimeframes = allTimeframes, subscribe } = module
  return { query, subscribe, supportedTimeframes }
}

export function getLowestTimeframe(supportedTimeframes: Timeframe[]) {
  return supportedTimeframes.find((x) => x !== "Block") as Timeframe
}
