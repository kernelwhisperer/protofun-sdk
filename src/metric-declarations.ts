import { PriceUnit, Timeframe } from "./primitives"
import { allTimeframes } from "./utils"

const METRIC_DECLARATIONS = {
  aave: [],
  btc: ["btc_price"],
  comp: ["tvl", "comp_price"],
  eth: ["base_fee", "eth_price", "tx_cost"],
  mkr: [],
} as const

/**
 * Protocols
 */
export type ProtocolId = keyof typeof METRIC_DECLARATIONS
export type Protocol = {
  enabled?: boolean
  id: ProtocolId
  title: string
  wip?: boolean
}

export const PROTOCOL_MAP: Record<ProtocolId, Protocol> = {
  aave: {
    id: "aave",
    title: "Aave",
  },
  btc: {
    enabled: true,
    id: "btc",
    title: "Bitcoin",
  },
  comp: {
    enabled: true,
    id: "comp",
    title: "Compound",
    wip: true,
  },
  eth: {
    enabled: true,
    id: "eth",
    title: "Ethereum",
  },
  mkr: {
    id: "mkr",
    title: "MakerDAO",
  },
} as const

export const PROTOCOL_IDS = Object.keys(PROTOCOL_MAP) as ProtocolId[]

export const PROTOCOLS = Object.values(PROTOCOL_MAP).sort((a, b) => {
  if (a.enabled === b.enabled) {
    return 0
  } else if (a.enabled && !b.enabled) {
    return -1
  } else {
    return 1
  }
})

/**
 * Metrics
 */

// Not working:
// type ValueOf<T> = T[keyof T];
// export type MetricId = ValueOf<typeof METRIC_LIST_DEFINITION>[number];
export type MetricIdForProtocol<T extends ProtocolId> = typeof METRIC_DECLARATIONS[T][number]
export type MetricId = MetricIdForProtocol<ProtocolId>

export type Variant = {
  label: string
  precision: number
  value?: string
}

export type Metric = {
  allowCompactPriceScale?: boolean
  disallowAlerts?: boolean
  disallowCandleType?: boolean
  disallowLiveMode?: boolean
  id: MetricId
  precision: number
  preferredLogScale?: boolean
  priceUnits: PriceUnit[]
  protocol: ProtocolId
  significantDigits: number[]
  timeframes: Timeframe[]
  title: string
  variants?: Variant[]
  wip?: boolean
}

// type MapType<T extends ProtocolId> = Record<
//   T,
//   Record<MetricIdsOfProtocol<T>, Metric>
// >;
type MetricsMapType = Record<ProtocolId, Partial<Record<MetricId, Metric>>>

export const METRICS_MAP: MetricsMapType = {
  aave: {},
  btc: {
    btc_price: {
      id: "btc_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "btc",
      significantDigits: [0],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "BTC price",
    },
  },
  comp: {
    comp_price: {
      id: "comp_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Comp price",
    },
    tvl: {
      allowCompactPriceScale: true,
      disallowCandleType: true,
      id: "tvl",
      precision: 1,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [0],
      timeframes: ["Hour", "Day"],
      title: "Total value locked",
      variants: [{ label: "V3 USDC - Wrapped Ether Pool", precision: 1 }],
      wip: true,
    },
  },
  eth: {
    base_fee: {
      id: "base_fee",
      precision: 1e9,
      preferredLogScale: true,
      priceUnits: [PriceUnit.GWEI],
      protocol: "eth",
      significantDigits: [2],
      timeframes: allTimeframes,
      title: "Base fee per gas",
    },
    eth_price: {
      id: "eth_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "eth",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Ether price",
    },
    tx_cost: {
      id: "tx_cost",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD, PriceUnit.ETH],
      protocol: "eth",
      significantDigits: [2, 5],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Transaction cost",
      variants: [
        { label: "ETH Transfer", precision: 1e18 / 21_000 },
        { label: "ERC20 Approval", precision: 1e18 / 45_000 },
        { label: "ERC20 Transfer", precision: 1e18 / 65_000 },
        { label: "NFT Transfer", precision: 1e18 / 85_000 },
        { label: "Uniswap V2 Swap", precision: 1e18 / 150_000 },
        { label: "Uniswap V3 Swap", precision: 1e18 / 185_000 },
        { label: "L2 Deposit", precision: 1e18 / 250_000 },
      ],
    },
  },
  mkr: {},
}

export const METRICS = PROTOCOL_IDS.map((protocolId) =>
  Object.values(METRICS_MAP[protocolId] || {})
).flat()

export function getMetric(protocolId: ProtocolId, metricId: MetricId): Metric {
  const metric = METRICS_MAP[protocolId][metricId]

  if (!metric) {
    throw new Error("Metric not found.")
  }

  return metric
}

export function getMetricPrecision(metric: Metric, variantIndex: number): number {
  const { variants, precision: defaultPrecision } = metric
  return variants ? variants[variantIndex].precision : defaultPrecision
}

export function getSignificantDigits(metric: Metric, priceUnitIndex: number): number {
  return metric.significantDigits[priceUnitIndex]
}
