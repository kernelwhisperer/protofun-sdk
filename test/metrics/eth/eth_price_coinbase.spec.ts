import { expect, it } from "vitest"

import query, { subscribe } from "../../../src/metrics/eth/eth_price"
import { wait } from "../../../src/utils"

it("Daily candles", async () => {
  // act
  const candles = await query({
    since: "1463529600",
    timeframe: "Day",
    until: "1464048000",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Daily candles - EIP 1559", async () => {
  // act
  const candles = await query({
    since: "1628121600",
    timeframe: "Day",
    until: "1628553600",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Hourly candles - EIP 1559", async () => {
  // act
  const candles = await query({
    since: "1628121600",
    timeframe: "Hour",
    until: "1628136000",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Block candles", async () => {
  // act
  try {
    await query({
      timeframe: "Block",
    })
  } catch (error) {
    // assert
    expect(error).toMatchInlineSnapshot(
      "[Error: Timeframe 'Block' is not supported for this metric.]"
    )
  }
})

it("Subscribe", async () => {
  // arrange
  const timeframe = "Minute"
  const candles = await query({
    limit: 1,
    timeframe,
  })
  // act
  const unsubscribe = subscribe({
    // eslint-disable-next-line no-console
    onError: console.error,
    onNewData: (data) => {
      candles.push(data)
      unsubscribe()
    },
    pollingInterval: 3_000,
    since: candles[0].timestamp,
    timeframe,
  })
  // assert
  await wait(5_000)
  expect(candles.length).to.equal(2)
})
