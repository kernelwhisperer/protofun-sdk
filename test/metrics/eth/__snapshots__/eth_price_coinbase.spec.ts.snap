// Vitest Snapshot v1

exports[`Daily candles - EIP 1559 1`] = `
[
  {
    "close": "2829.35",
    "high": "2845",
    "low": "2530.69",
    "open": "2724.71",
    "timestamp": "1628121600",
    "volume": "326176.98753449",
  },
  {
    "close": "2892.98",
    "high": "2949.21",
    "low": "2721",
    "open": "2829.49",
    "timestamp": "1628208000",
    "volume": "283374.42059615",
  },
  {
    "close": "3163.09",
    "high": "3173.72",
    "low": "2866.58",
    "open": "2892.83",
    "timestamp": "1628294400",
    "volume": "299359.36243726",
  },
  {
    "close": "3014.13",
    "high": "3191.99",
    "low": "2948.76",
    "open": "3163.09",
    "timestamp": "1628380800",
    "volume": "244654.45876056",
  },
  {
    "close": "3162.34",
    "high": "3190",
    "low": "2892.5",
    "open": "3014.12",
    "timestamp": "1628467200",
    "volume": "315853.32665056",
  },
  {
    "close": "3142.09",
    "high": "3235.1",
    "low": "3051",
    "open": "3162.33",
    "timestamp": "1628553600",
    "volume": "223500.03188943",
  },
]
`;

exports[`Daily candles 1`] = `
[
  {
    "close": "13.18",
    "high": "14.93",
    "low": "12.5",
    "open": "12.5",
    "timestamp": "1463529600",
    "volume": "482.52182654",
  },
  {
    "close": "14.9",
    "high": "14.9",
    "low": "13",
    "open": "13.18",
    "timestamp": "1463616000",
    "volume": "950.44120523",
  },
  {
    "close": "14.17",
    "high": "14.82",
    "low": "13.71",
    "open": "14.82",
    "timestamp": "1463702400",
    "volume": "254.16661711",
  },
  {
    "close": "13.61",
    "high": "13.91",
    "low": "13.61",
    "open": "13.86",
    "timestamp": "1463961600",
    "volume": "180.78671194",
  },
  {
    "close": "12.77",
    "high": "13.74",
    "low": "12",
    "open": "13.68",
    "timestamp": "1464048000",
    "volume": "2753.23929761",
  },
]
`;

exports[`Hourly candles - EIP 1559 1`] = `
[
  {
    "close": "2705.07",
    "high": "2742",
    "low": "2697.11",
    "open": "2724.71",
    "timestamp": "1628121600",
    "volume": "7799.42127091",
  },
  {
    "close": "2690.02",
    "high": "2708.78",
    "low": "2665",
    "open": "2705.07",
    "timestamp": "1628125200",
    "volume": "12385.58325474",
  },
  {
    "close": "2698.39",
    "high": "2700",
    "low": "2675",
    "open": "2690.01",
    "timestamp": "1628128800",
    "volume": "4633.67463858",
  },
  {
    "close": "2712.12",
    "high": "2721.16",
    "low": "2690.66",
    "open": "2698.39",
    "timestamp": "1628132400",
    "volume": "7160.95978414",
  },
  {
    "close": "2694.42",
    "high": "2721.1",
    "low": "2679",
    "open": "2711.98",
    "timestamp": "1628136000",
    "volume": "5054.41449922",
  },
]
`;
